[Merged by Bors] - feat(Probability/Kernel): disintegration of finite kernels

RemyDegenne opened the pull request:
Let `κ : kernel α (β × Ω)` be a finite kernel, where `Ω` is a standard Borel space. Then if `α` is countable or `β` has a countably generated σ-algebra (for example if it is standard Borel), then there exists a `kernel (α × β) Ω` called conditional kernel and denoted by `condKernel κ` such that `κ = fst κ ⊗ₖ condKernel κ`.

Properties of integrals involving `condKernel` are collated in the file `Integral.lean`.
The conditional kernel is unique (almost everywhere w.r.t. `fst κ`): this is proved in the file `Unique.lean`.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sgouezel commented:
When alpha is countable, can you explain why you need `MeasurableSingletonClass`? Otherwise, I would expect that everything is constant on the atoms of the partition generated by the sigma-algebra, and the quotient space is countable, so the general case should be reducable to the `MeasurableSingletonClass` case. I'm absolutely not saying that it should be implemented in this way with a quotient, rather that with this heuristics I don't see mathematically why `MeasurableSingletonClass` should be relevant.

RemyDegenne commented:
The non-mathematical reason I added that hypothesis is that I used `measurable_from_prod_countable` (in the file `Probability.Kernel.Disintegration.Basic`). It asked for that typeclass and I added it without thinking about it.

RemyDegenne commented:
I don't see how I could remove the `MeasurableSingletonClass` assumption without adding another assumption somewhere. To conclude that everything is constant on the atoms, I would need the codomain to have measurable singletons? Otherwise the function could take several values in a non-singleton atom, as long as all those values belong to some atom of the codomain?

Also the existence of a partition of atoms would mean that the measurable space is countably generated, but I think that's not necessarily the case even if the space is countable.

sgouezel commented:
> I don't see how I could remove the `MeasurableSingletonClass` assumption without adding another assumption somewhere. To conclude that everything is constant on the atoms, I would need the codomain to have measurable singletons? Otherwise the function could take several values in a non-singleton atom, as long as all those values belong to some atom of the codomain?

I would think that everything is done taking preimages of measurable sets under measurable maps. Such preimages should be constant over atoms, right? (Warning: I haven't looked at the code, so I might be completely wrong)

> Also the existence of a partition of atoms would mean that the measurable space is countably generated, but I think that's not necessarily the case even if the space is countable.

I don't see the problem here. Define the atom of `x` as the intersection of all measurable sets that contain `x`. For any `y` not in the atom, one can take a measurable set containing `x` but not `y`, so the atom is in fact a countable intersection (of such sets over all `y` not belonging to the atom), and it belongs to the sigma-algebra. Then any measurable set is the (countable) union of the atoms it contains. But again, I don't think one should go through this argument for the construction, rather see where `MeasurableSingletonClass` is used and if one can remove it locally. If you want, I can have a look in the next days.

RemyDegenne commented:
@sgouezel All prerequisites are now merged (thanks for the reviews!), and this is the final disintegration PR. I did not resolve your question about the `MeasurableSetSingleton` hypothesis: if you have time to have a look at the construction and see if you can remove that hypothesis, that would be great!

sgouezel commented:
I've just pushed a commit removing `MeasurableSetSingleton`, following the sketch above. Can you tell me if you're happy with the modification? If so, I'll review the rest of the PR.

RemyDegenne commented on Mathlib/Probability/Kernel/Disintegration/Basic.lean:
```diff
@@ -0,0 +1,459 @@
+    refine (measurable_from_prod_countable' (fun a ↦ ?_) ?_).comp measurable_swap
+    · exact kernel.measurable (κ a).condKernel
+    · intro y y' x hy'
+      have : κ y' = κ y := by
+        ext s hs
+        exact mem_of_mem_measurableAtom hy'
+          (kernel.measurable_coe κ hs (measurableSet_singleton (κ y s))) rfl
```
@sgouezel That's the part of your explanation I did not quite get before. Now I see it. Thanks a lot!

RemyDegenne commented:
Thanks for the help! Your code looks good. I saw a place where I think I can refactor my code to improve its clarity, so I'm turning that PR back to awaiting-author for now and will come back to you when it's again ready for review.

RemyDegenne commented:
Actually I realize that I had already tried that refactor and it does not bring much benefits, so the PR is ready for review!

More explanations about that:
The definition `condKernelBorelSnd` takes a conditional CDF `f` (on the reals) and builds a kernel on a Borel space by going to the reals, building a kernel from the CDF (`hf.toKernel f`), then going back to the Borel space. It looks like I could first introduce another definition for the conditional kernel on R, then have `condKernelBorelSnd` do only the map and comap.
But actually `condKernelBorelSnd` does something slightly more subtle than map-then-comap and needs access to more properties of the kernel built from a conditional CDF (notably `ae_toKernel_eq_one` to show the conditional property), such that it is quite specific to the CDF construction and I might as well dispense with the intermediate definitions.

sgouezel commented:
I'm confused by github here. I got an email saying that you had realized that the refactor wasn't worth it, and that the PR was ready for review, but I can't see your message on github and the PR is still awaiting author. Can you confirm that I did not dream this message, and that the PR is indeed ready for review?

RemyDegenne commented:
The PR is not ready for review.
I wrote that message too fast, changed my mind once again about the refactor and then deleted the message. I forgot that github also sends an email.
The refactor is in progress. Sorry for the confusion!

RemyDegenne commented:
The refactor is done and this is ready for review. The PR is big, but if I dismantle the old disintegration file I have to have all the replacements in the PR (which means I can't PR the files about integrals and uniqueness later).

sgouezel commented on Mathlib/Probability/Kernel/Disintegration/Basic.lean:
```diff
@@ -0,0 +1,518 @@
+
+open ProbabilityTheory.kernel
+
+lemma isRatCondKernelCDFAux_density_Iic (κ : kernel α (γ × ℝ)) [IsFiniteKernel κ] :
```
The next 70 lines are about kernels with values in a product where the second component is the real line, so they're interrupting a little bit the flow of the argument where you're moving things from the real line to a general standard Borel space. Would it make sense to move them to the top of the file (before the `BorelSnd` section)? Or to `CdfToKernel.lean`?

sgouezel commented on Mathlib/Probability/Kernel/Disintegration/Basic.lean:
```diff
@@ -0,0 +1,518 @@
+/-- Conditional kernel of a measure on a product space: a Markov kernel such that
+`ρ = ρ.fst ⊗ₘ ρ.condKernel` (see `MeasureTheory.Measure.compProd_fst_condKernel`). -/
+noncomputable
+def _root_.MeasureTheory.Measure.condKernel (ρ : Measure (α × Ω)) [IsFiniteMeasure ρ] :
```
`irreducible_def`, maybe, to hide the messy details of the construction (which should never be relevant once the API is set up)?

sgouezel commented on Mathlib/Probability/Kernel/Disintegration/Integral.lean:
```diff
@@ -0,0 +1,280 @@
+section Integral
+
+variable {ρ : Measure (β × Ω)} [IsFiniteMeasure ρ]
+  {E : Type*} {f : β × Ω → E} [NormedAddCommGroup E] [NormedSpace ℝ E] [CompleteSpace E]
```
It seems to me that the `[CompleteSpace E]` assumption shouldn't be needed: if the space is not complete, then all integrals vanish by our choice of junk values, so all the theorems below remain true. I've taken the liberty to push a commit removing some of these assumptions here and there.

RemyDegenne replied:
Great, thanks!

sgouezel commented on Mathlib/Probability/Kernel/Disintegration/Unique.lean:
```diff
@@ -0,0 +1,178 @@
+This theorem in the case of finite kernels is weaker than `eq_condKernel_of_measure_eq_compProd`
+which asserts that the kernels are equal almost everywhere and not just on a given measurable
+set. -/
+theorem eq_condKernel_of_measure_eq_compProd' (κ : kernel α Ω) [IsSFiniteKernel κ]
```
I think I am unhappy with the assumptions in this theorem and in the next ones: if `ρ = ρ.fst ⊗ₘ κ` and `ρ` is a finite measure, then `κ a` should have finite mass almost everywhere wrt `ρ.fst`, and I feel that this should be sufficient to carry out the arguments in the lemma without the assumptions of s-finiteness or of finiteness. However, the lemmas were already there (at least in the measure version) and you just copied them, so this shouldn't hold the PR back. I'll investigate that once this one is merged, if you agree.

sgouezel reviewed (commented):
This looks great, thanks! Sorry for the delay to review the PR (it means there is another merge conflict that you have to solve...)
bors d+

RemyDegenne commented:
Thanks for all the reviews and for your help on the disintegration PRs. No worries about the review time: I don't expect a quick review for a 1000 lines PR :)
I agree that the assumptions in the Unique file don't feel right. I won't work on that myself for now (at least not for the next days), so feel free to have a look.

bors r+
